_posts/2021-03-10-dask_memory_usage.md

```diff
@@ -0,0 +1,261 @@
+"('from_sequence-3637e6ff937ef8488894df60a80f62ed', 0)",51.70703125,51.70703125
+"('from_sequence-3637e6ff937ef8488894df60a80f62ed', 1)",51.28125,51.78515625
+"('from_sequence-3637e6ff937ef8488894df60a80f62ed', 2)",51.30859375,51.30859375
+"('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 2)",56.19140625,56.19140625
+"('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 0)",51.70703125,54.26953125
+"('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 1)",52.30078125,52.30078125
+"('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 3)",51.48046875,384.00390625
+```
+
+For each task in the graph we are told minimum memory usage and peak memory usage, in MB.
+
+In more readable form:
+
+| task_key                                                   | min_memory_mb | max_memory_mb |
+| "('from_sequence-3637e6ff937ef8488894df60a80f62ed', 3)"    |    51.2421875 |    51.2421875 |
+| "('from_sequence-3637e6ff937ef8488894df60a80f62ed', 0)"    |   51.70703125 |   51.70703125 |
+| "('from_sequence-3637e6ff937ef8488894df60a80f62ed', 1)"    |      51.28125 |   51.78515625 |
+| "('from_sequence-3637e6ff937ef8488894df60a80f62ed', 2)"    |   51.30859375 |   51.30859375 |
+| "('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 2)" |   56.19140625 |   56.19140625 |
+| "('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 0)" |   51.70703125 |   54.26953125 |
+| "('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 1)" |   52.30078125 |   52.30078125 |
+| "('calculate_top_10-afc867e38c3bd0aac8c18bb00d3634ca', 3)" |   51.48046875 |  384.00390625 |
+
+The bottom four lines are the interesting ones; all four start with a minimum memory usage of ~50MB RAM, and then perhaps memory increases as the code runs.
+How much it increases presumably depends on the size of the files; most of them are quite small, so memory usage doesn't change much.
+**One file uses much more maximum memory than the others, 384MB of RAM; presumably it's `big.txt` which is 25MB, since the other files are all smaller than 1MB.**
+
+The mechanism used, polling peak process memory, has some limitations:
+
+* You'll notice there's a `gc.collect()` at the top of the `calculate_top_10()`; this ensures we don't count memory from previous code that hasn't been cleaned up yet.
+* There's also a `sleep()` at the bottom of `calculate_top_10()`.
+  Because polling is used, tasks that run too quickly won't get accurate information.
+* Finally, because polling is per-process, you can't run multiple threads per worker, otherwise you'll get results that combine memory usage of multiple tasks.
+
+## Interpreting the data
+
+What we've learned is that memory usage of `calculate_top_10()` grows with file size; this can let us [characterize the memory requirements for the workload](https://pythonspeed.com/articles/estimating-memory-usage/), creating a model saying "for these file sizes and this level of parallelism we will need this much RAM."
```

Comment:
The second half of this sentence is a little clunky. It might be clearer if there was a full stop after the link "characterize the memory requirements for the workload". Then we could rephrase the next part, and get rid of the quotes.

BTW, I use quotes like this all the time when I'm writing informally, but it's generally a warning sign that I could rephrase things in a clearer way.